controller/network/routesender.go

```diff
@@ -0,0 +1,168 @@
+}
+
+func (self *routeSenderController) removeRouteSender(rs *routeSender) {
+	self.senders.Remove(rs.sessionId)
+}
+
+type routeSender struct {
+	sessionId  string
+	circuit    *Circuit
+	routeMsgs  []*ctrl_pb.Route
+	timeout    time.Duration
+	in         chan *routeStatus
+	attendance map[string]bool
+}
+
+func newRouteSender(sessionId string, timeout time.Duration) *routeSender {
+	return &routeSender{
+		sessionId:  sessionId,
+		timeout:    timeout,
+		in:         make(chan *routeStatus, 16),
+		attendance: make(map[string]bool),
+	}
+}
+
+func (self *routeSender) route(attempt uint32, circuit *Circuit, routeMsgs []*ctrl_pb.Route, strategy xt.Strategy, terminator xt.Terminator) (peerData xt.PeerData, cleanups map[string]struct{}, err error) {
+	// send route messages
+	tr := circuit.Path[len(circuit.Path)-1]
+	for i := 0; i < len(circuit.Path); i++ {
+		r := circuit.Path[i]
+		go self.sendRoute(r, routeMsgs[i])
+		self.attendance[r.Id] = false
+	}
+
+	deadline := time.Now().Add(self.timeout)
+	timeout := time.Until(deadline)
+attendance:
+	for {
+		select {
+		case status := <-self.in:
+			if status.success {
+				if status.attempt == attempt {
+					logrus.Debugf("received successful route status from [r/%s] for attempt [#%d] of [s/%s]", status.r.Id, status.attempt, status.sessionId)
+
+					self.attendance[status.r.Id] = true
+					if status.r == tr {
+						peerData = status.peerData
+						strategy.NotifyEvent(xt.NewDialSucceeded(terminator))
+					}
+				} else {
+					logrus.Warnf("received successful route status from [r/%s] for alien attempt [#%d (not #%d)] of [s/%s]", status.r.Id, status.attempt, attempt, status.sessionId)
+				}
+
+			} else {
+				if status.attempt == attempt {
+					logrus.Warnf("received failed route status from [r/%s] for attempt [#%d] of [s/%s] (%v)", status.r.Id, status.attempt, status.sessionId, status.rerr)
+
+					if status.r == tr {
+						strategy.NotifyEvent(xt.NewDialFailedEvent(terminator))
+					}
+					cleanups = self.cleanups(circuit)
+
+					return nil, cleanups, errors.Errorf("error creating route for [s/%s] on [r/%s] (%v)", self.sessionId, status.r.Id, status.rerr)
+				} else {
+					logrus.Warnf("received failed route status from [r/%s] for alien attempt [#%d (not #%d)] of [s/%s]", status.r.Id, status.attempt, attempt, status.sessionId)
+				}
+			}
+
+		case <-time.After(timeout):
+			cleanups = self.cleanups(circuit)
+			return nil, cleanups, errors.Errorf("timeout creating routes for [s/%s]", self.sessionId)
+		}
+
+		allPresent := true
+		for _, v := range self.attendance {
+			if !v {
+				allPresent = false
+			}
+		}
+		if allPresent {
+			break attendance
+		}
+
+		timeout = time.Until(deadline)
+	}
+
+	return peerData, nil, nil
+}
+
+func (self *routeSender) sendRoute(r *Router, routeMsg *ctrl_pb.Route) {
+	body, err := proto.Marshal(routeMsg)
+	if err != nil {
+		logrus.Errorf("error marshalling route message for [s/%s] to [r/%s] (%v)", routeMsg.SessionId, r.Id, err)
+		return
+	}
```

Comment:
Should we fast fail here if the control channel is closed and returns an error

Reply:
Short answer: We can't, it's happning in a goroutine.

Long answer: We could, but this shouldn't happen, and if it does, it will get timed out. Implementing fail fast plumbing for this from a goroutine would add a lot of convolution for little gain, imo.